### DISCUSSION TITLE
`error.js` can NOT catch error from async function when using App Router!

### DISCUSSION THREAD

**Discussion by jkiss** [vercel/next.js#50564] [Help] [answered]
### Summary

```js
const Story = ({ params: { slug } }: Props) => {
    const doInitData = async () => {
        throw new Error("test")
    }

    useEffect(() => {
        doInitData()
    }, [])

    return <>...</>
}

export default Story
```

The above code throwing error can NOT be catched by `error.js`, but below code can: 👇

```js
const Story = ({ params: { slug } }: Props) => {
    const doInitData = async () => {
        throw new Error("test")
    }

    useEffect(() => {
        throw new Error("test")

        // doInitData()
    }, [])

    return <>...</>
}

export default Story
```

Why? really need help, how `error.js` catch error from `async` function???

**Comment by devjiwonchoi** [comment, marked as answer]
Since async functions could exist and be called inside the `useEffect` hook, the `Error` will be thrown successfully.

The issue is on the `Error Boundary`.

`error.js` is based on [React Error Boundary](https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary).

> Automatically wrap a route segment and its nested children in a [React Error Boundary](https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary).

In React Error Boundary, there are a few cases that the error boundaries cannot catch errors.

> Error boundaries do not catch errors for:
> - Event handlers (learn more)
> - Asynchronous code (e.g. setTimeout or requestAnimationFrame callbacks)
> - Server side rendering
> - Errors thrown in the error boundary itself (rather than its children)

Ref: https://legacy.reactjs.org/docs/error-boundaries.html

As you can see, `asynchronous code` escape from the error boundaries, which is exactly your case.

Fortunately, there's a way to resolve this issue by using `useState` to re-throw error inside the `useEffect`.

```javascript
'use client';
import { useEffect, useState } from 'react';

// type Props...

const Story = ({ params: { slug } }: Props) => {
  const [, setError] = useState();

  const doInitData = async () => {
    throw new Error('test');
  };

  useEffect(() => {
    const load = async () => {
      try {
        await doInitData();
      } catch (err) {
        setError(() => {
          throw err;
        });
      }
    };

    load();
  }, []);

  return <>...</>
}

export default Story;
```

Ref: https://stackoverflow.com/questions/67889383/react-error-boundaries-with-useeffect-and-async-function-what-im-missing

In summary, the `Error` can be successfully thrown by calling the `doInitData` inside the `useEffect`,
but the `Error Boundary` cannot catch `async codes` resulting in not rendering the `error.js`.

Therefore you can resolve this issue by re-throwing the error inside the `useEffect`.

**Reply by singleseeker** [reply]
Hi, what should we do if we use an Event handlers?

**Reply by jkiss** [reply]
in the same way? actually event handlers is also a async function

**Reply by Fiyin-Anne** [reply]
Hi, pls how can I apply this on a submit action?

**Reply by alvieridev** [reply]
A submit action is equally an async process

**Reply by thekevinkun** [reply]
Hi, your solution is work successfully! But I think it also happen for `loading.js`. It's also not rendering `loading.js` when I fetch something inside `useEffect`. I think it's the same problem as `error.jsx`. Do you have a solution for this?
